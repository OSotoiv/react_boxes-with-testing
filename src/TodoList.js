import React, { useState } from "react";
import { v4 as uuidv4 } from 'uuid';
import TodoForm from './TodoForm';
import Todo from './Todo'

const INIT_FORM_STATE = { todo: "" }
const TodoList = () => {
    const [listOfTodos, setlistOfTodos] = useState([])
    const [formData, setFormData] = useState(INIT_FORM_STATE)
    function updateFormState(e) {
        const { name, value } = e.target
        setFormData(formDate => ({ ...formDate, [name]: value }))
    }
    function addTodo(data) {
        //make new box then update the list of boxes to new list with new box added
        const newTodo = { ...data, id: uuidv4(), state: 'incomplete' }
        setlistOfTodos([...listOfTodos, newTodo]);

    }
    function submitTodoForm(e) {
        //when form submits create a new todo with form data
        e.preventDefault()
        addTodo(formData);
        setFormData(INIT_FORM_STATE);
    }
    function removeTodo(id) {
        setlistOfTodos(listOfTodos.filter(tood => (tood.id !== id)))
    }
    function isDone(id) {
        setlistOfTodos(listOfTodos.map(listItem => {
            console.log(listItem)
            console.log(id)
            if (listItem.id === id) {
                listItem.state = 'done'
            }
            return listItem;
        }))
    }
    return (<>
        <TodoForm formState={formData} submitTodoForm={submitTodoForm} updateFormState={updateFormState} />
        <ul>
            {listOfTodos.map(todo => <Todo task={todo} id={todo.id} key={todo.id} removeTodo={removeTodo} isDone={isDone} />)}
        </ul>
    </>)
}
export default TodoList;