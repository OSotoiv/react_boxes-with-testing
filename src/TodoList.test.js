import React from "react";
import TodoList from './TodoList'
import { render, fireEvent } from "@testing-library/react";

it('should pass SMOKE TEST', () => {
    render(<TodoList />)
});
it('matches snapshot', () => {
    const { asFragment } = render(<TodoList />)
    expect(asFragment()).toMatchSnapshot();
});

it('should render a Form to add a Todo', () => {
    const { getByLabelText, getByRole } = render(<TodoList />);
    const todoFormInput = getByLabelText('Todo:');
    expect(todoFormInput).toBeInTheDocument();
    expect(todoFormInput.value).toEqual("");
    expect(getByRole('button')).toBeInTheDocument();
});
it('should start with no todos', () => {
    const { queryByTestId } = render(<TodoList />);
    expect(queryByTestId('todo')).toBeNull()
});
it('should render a todo when form is submitted', () => {
    const { queryByTestId, getByLabelText, getByRole } = render(<TodoList />);
    //no todos when app starts
    expect(queryByTestId('todo')).toBeNull()

    //submit form for a new todo
    const todoFormInput = getByLabelText('Todo:');
    fireEvent.change(todoFormInput, { target: { value: 200 } });
    const todoFormButton = queryByTestId('submit-todo-btn');
    fireEvent.click(todoFormButton);

    //now todo should be on the page
    const todoItem = queryByTestId('todo')
    expect(todoItem).not.toHaveStyle('textDecoration: line-through');

    //should strike through todo when done button clicked
    const li = queryByTestId('todo');
    fireEvent.mouseEnter(li)
    const doneBtn = queryByTestId('is-done')
    fireEvent.click(doneBtn)
    expect(queryByTestId('todo')).toHaveStyle('textDecoration: line-through')

    const removeBtn = queryByTestId('todo-remove');
    fireEvent.click(removeBtn);
    expect(queryByTestId('todo')).toBeNull()

});
// it('should remove a todo when the `remove` button clicked')
// it('should strike through the todo when the `done` button is clicked')