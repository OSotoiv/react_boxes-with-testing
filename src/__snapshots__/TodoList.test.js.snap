// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="todo"
    >
      Todo:
    </label>
    <input
      id="todo"
      name="todo"
      placeholder="Enter todo"
      required=""
      type="text"
      value=""
    />
    <button
      data-testid="submit-todo-btn"
      type="submit"
    >
      Submit
    </button>
  </form>
  <ul />
</DocumentFragment>
`;
